### PR TITLE
Rough case tile template preview

### PR DESCRIPTION
**Pull request by orangejenny**
Opening for feedback on the general idea.

This is just an idea, something to give app builders a better idea of what they're getting into with a case tile template. The down side is that a developer updating a template would also need to update the HTML preview, which is a bit annoying. This would also be less relevant if we stick to [standard guidelines](https://github.com/dimagi/commcare-hq/pull/33128/files#diff-c70643f3bf56157db628ae1b704ccd5e409b9436163a74deff8e61070f85d408R7) for naming templates and fields.

Screen Shot 2023-06-19 at 5 31 56 PM

## Feature Flag
REC case tile

### Safety story
Additive change to an app builder feature flag not especially widely used.

### Automated test coverage

No

### QA Plan

Not requesting QA

### Rollback instructions

- [x] This PR can be reverted after deploy with no further considerations

### Labels & Review
- [x] Risk label is set correctly
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change

### CODE REVIEW

**Review by snopoke — COMMENTED** [review]
Can we not autogenerate the preview?

If we store the template config as a JSON object I think we could probably generate the tile XML as well.

For example, I think this would be sufficient to generate the [one_two_one.xml](https://github.com/dimagi/commcare-hq/pull/33128/files#diff-3f7253dbb806f92c0871d33f4a6c59c6d5ce3b26d5171def00b6dc9fb0b4f666) template:
```json
{
  "fields": [
    {"name": "header", "grid": {"height": 1, "width": 12, "x": 0, "y": 0}, "type": "plain"},
    {"name": "map", "grid": {"height": 1, "width": 6, "x": 6, "y": 0}, "type": "map"},
    {"name": "top_left", "grid": {"height": 1, "width": 7, "x": 0, "y": 1}, "type": "plain"},
    {"name": "top_right", "grid": {"height": 1, "width": 4, "x": 8, "y": 1}, "type": "plain"},
    {"name": "bottom", "grid": {"height": 1, "width": 12, "x": 0, "y": 2}, "type": "plain"},
  ]
}
```

**Comment by MartinRiese** [comment]
I think a proper naming convention would be good enough. Auto generating sounds nice but then you have things like the map field that does not show up in the grid.

**Comment by esoergel** [comment]
I love the idea, I was thinking similarly when working on this yesterday - it would be very nice to represent this visually in some way.  No strong feelings on how to define it, though it would be nice to autogenerate if we're pretty confident we're mostly concerned about layout, not styling.

It would be cool if we could also encode the layouts in a pseudo-graphical representation, eg, for one_two_one.txt:

```
aaaaaaaaaaaa
bbbbbbb cccc
dddddddddddd
```
Though I wouldn't want to deal with the code necessary to turn that into an HTML or XML layout.  That'd be a good dev interview type coding puzzle, I bet.

**Comment by orangejenny** [comment]
Cleaned this up a bit, looking to actually merge it. It's a small step towards allowing app builders to create their own templates in the UI. Could I have another round of review?

I switched from HTML to a json representation of just the grid. It's admittedly duplicative with the tile XML. We'd need a more detailed json structure to automatically generate the tile (for invisible fields and attributes like horz-align and vert-align). I didn't get into that. The preview does leave off invisible fields - I thought about listing them out but decided that wouldn't add much value.

**Review by MartinRiese — COMMENTED** [review]
Could you add a note to the readme in the case_tile_templates directory about keeping them coordinated in sync with the actual templates?

File: corehq/apps/app_manager/suite_xml/features/case_tiles.py [hunk 1]
```diff
@@ -41,11 +42,14 @@ def filepath(self):
...
-        encoding='utf-8'
-    ) as f:
-        data = json.loads(f.read())
+    try:
+        with open(
+            TILE_DIR / (template + '.json'),
+            encoding='utf-8'
+        ) as f:
+            data = json.loads(f.read())
+    except FileNotFoundError:
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/suite_xml/features/case_tiles.py:51]: Without the check the page 500s, which is too severe.

File: corehq/apps/app_manager/suite_xml/case_tile_templates/one_two_one_one.json [hunk 2]
```diff
@@ -1,5 +1,13 @@
 {
     "slug": "one_two_one_one",
     "filename": "one_two_one_one.xml",
-    "fields": ["header", "map", "map_popup", "top_left", "top_right", "middle", "bottom"]
+    "fields": ["header", "map", "top_left", "top_right", "middle", "bottom"],
+    "fields": ["header", "map", "map_popup", "top_left", "top_right", "middle", "bottom"],
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/suite_xml/case_tile_templates/one_two_one_one.json:4]: nope - 9693cdca555

File: corehq/apps/app_manager/suite_xml/features/case_tiles.py [hunk 3]
```diff
@@ -41,11 +42,14 @@ def filepath(self):
...
+        data = {}
```
[the remark this answers is not in the document]
Comment (orangejenny) [reply corehq/apps/app_manager/suite_xml/features/case_tiles.py:52]: The page loads, but the case tile mapping dropdowns are blank and there's no preview.

**Review by MartinRiese — APPROVED** [review]

**Pull request merged**